### PR TITLE
x509pop server plugin support for servers trust bundle

### PR DESCRIPTION
**Pull request by kfox1111**
Enables the x509pop node attestor server plugin to be configured to use the SPIRE Servers own trust bundle.

### CODE REVIEW

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 1]
```diff
@@ -97,14 +106,51 @@ type Plugin struct {
...
 	return &Plugin{}
 }
 
+func (p *Plugin) BrokerHostServices(broker pluginsdk.ServiceBroker) error {
+	if !broker.BrokerClient(&p.identityProvider) {
+		return status.Errorf(codes.FailedPrecondition, "IdentityProvider host service is required")
+	}
+	go func() {
+		ctx := context.Background()
+		sleep := 1 * time.Second
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:125]: Rather than kick off a goroutine that may or may not be run after the IdentityProvider host service is ready, it would be better to lazily initialize this in the first call to Attest.

**Comment by kfox1111** [comment]
> Using the bundle, without any sort of restriction, seems scary. Doesn't this imply that any workload can just turn around and attest as a node?

Yeah, that I think can be handled like:
```
    NodeAttestor "x509pop" {
      plugin_data {
        spire_trust_bundle = true
        # Only allow spiffe identifiers of the form spiffe://<trustDomain>/k8s-spire-agent-helper/<hostname>
        # Generate identifiers that look like spiffe://<trustDomain>/x509pop/k8s-spire-agent/<hostname>
        agent_path_template = "{{ ${d}p := printf \"spiffe://%s/k8s-spire-agent-helper/\" .TrustBundle }}{{ ${d}s := printf \"%s\" (index .Certificate.URIs 0) }}{{ if gt (len ${d}s) (len ${d}p) }}{{ ${d}ps := slice ${d}s 0 (len ${d}p)
 }}{{ if eq ${d}ps ${d}p }}{{ printf \"/x509pop/k8s-spire-agent/%s\" (slice ${d}s (len ${d}p)) }}{{ end }}{{ end }}"
      }
    }
```

if it agent_path renders out to "", then it should be disallowed.

I did hit:
* https://github.com/spiffe/spire/issues/5574
* https://github.com/spiffe/spire/issues/5573

Trying to get it to work though.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go (hunk 1 above)
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go]: Done

**Comment by azdagron** [comment]
Hey @kfox1111, do you have an arch diagram or something that spells out the use case clearly that you can share (either here or privately in slack with the maintainer group)? Before we consider taking this we'd like to see if there are alternatives.

**Comment by kfox1111** [comment]
Here's what I have done / working on doing. Got the host -> k8s part working. Still working on the k8s -> vm part.
x509pop

**Comment by kfox1111** [comment]
Related to: https://github.com/spiffe/spire/issues/5206

**Comment by kfox1111** [comment]
Used with the sprig pr, this works:

        agent_path_template = "{{ $$p := printf \"spiffe://%s/k8s-spire-agent-helper/\" .TrustDomain }}{{ $$s := printf \"%s\" (index .Certificate.URIs 0) }}{{ if hasPrefix $$p $$s }}{{ printf \"/x509pop/k8s-spire-agent/%s\" (trimPrefix $$p $$s) }}{{ else }}{{ fail \"Invalid prefix\" }}{{ end }}"

**Comment by kfox1111** [comment]
Updated diagram:
x509pop

**Comment by kfox1111** [comment]
@sorindumitru I think we're pretty much on the same page.

We could make a new x509svidpop plugin but it would end up being almost the same code as the x509pop one. (The patch here is pretty small). So a lot more stuff to maintain.

the svid use case really isn't that different from the normal x509pop one. Either way you probably want some way to filter out which certs are allowed to a subset of all possible ones. That can be done via agent_path_template as done in this pr without any special code in the plugin.

100% agree on the future ability to use some other spire trustDomain too. Was going to propose that in a different patch. The primary use case for me being:

k8s multitenant cluster. Secured with spire. Call this, the resource provider cluster.

Tenant uses resource provider cluster with kubevirt to launch a bunch of vm's for their tenant. Inside, they launch their own spire server, and workloads (maybe even their own k8s cluster in the vms). It would greatly simplify management if they could use x509 certs from the resource provider cluster to node attest to their own spire-server.

That again, would basically be this same x509pop plugin as described here, but getting its trustBundle from disk and keep it refreshed (existing functionality extended to refresh from disk), or get it over vsock (not sure which is best yet)

**Comment by sorindumitru** [comment]
I agree it's going to be very similar to `x509pop`, but dealing with that is an implementation detail, I'm sure some parts can be shared. I just don't think that the template language is a good way of dealing with those differences.

File: doc/plugin_server_nodeattestor_x509pop.md [hunk 2]
```diff
@@ -18,9 +18,11 @@ spiffe://<trust_domain>/spire/agent/x509pop/<fingerprint>
...
 |-----------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------|
+| `spire_trust_bundle`     | If true, use the spire servers own trust bundle to use for validation.                                                                                                                                                                      |                                         |
+| `svid_prefix`            | The prefix of the SVID to use for matching vaid SVIDS and exchanging them for Node SVIDs                                                                                                                                                    | /spire-exchange                         |
```
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:22]: Maybe `svid_path_prefix`? Also...
```suggestion
| `svid_prefix`            | The prefix of the SVID to use for matching valid SVIDS and exchanging them for Node SVIDs                                                                                                                                                    | /spire-exchange                         |
```

File: doc/plugin_server_nodeattestor_x509pop.md [hunk 3]
```diff
@@ -18,9 +18,11 @@ spiffe://<trust_domain>/spire/agent/x509pop/<fingerprint>
...
 | `ca_bundle_paths`     | A list of paths to trusted CA bundles on disk. The files must contain one or more PEM blocks forming the set of trusted root CA's for chain-of-trust verification.                                                                             |                                         |
-| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `"{{ .PluginName}}/{{ .Fingerprint }}"` |
+| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `See Agent Path Template for details`   |
```
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:25]: nit: use a section link to the correct section (https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#section-links)

File: doc/plugin_server_nodeattestor_x509pop.md [hunk 4]
```diff
@@ -43,9 +45,27 @@ A sample configuration:
 | SHA1 Fingerprint | `x509pop:ca:fingerprint:0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33` | The SHA1 fingerprint as a hex string for each cert in the PoP chain, excluding the leaf. |
 | SerialNumber     | `x509pop:serialnumber:0a1b2c3d4e5f`                               | The leaf certificate serial number as a lowercase hexadecimal string                     |
 
+## SVID Prefix
```
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:48]:
```suggestion
## SVID Path Prefix
```

File: doc/plugin_server_nodeattestor_x509pop.md [hunk 5]
```diff
@@ -57,3 +77,4 @@ Some useful values are:
 | .TrustDomain          | The configured trust domain                                                                  |
 | .Subject.CommonName   | The common name field of the agent's x509 certificate                                        |
 | .SerialNumberHex      | The serial number field of the agent's x509 certificate represented as lowercase hexadecimal |
+| .SVIDPath             | The SVID Path after removing the SVID Prefix                                                 |
```
Comment (azdagron) [inline doc/plugin_server_nodeattestor_x509pop.md:80]: Maybe `SVIDPathSuffix` better implies that a prefix has been stripped?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 6]
```diff
@@ -35,15 +38,19 @@ func builtin(p *Plugin) catalog.BuiltIn {
 }
 
 type Config struct {
+	SPIRETrustBundle  bool     `hcl:"spire_trust_bundle"`
+	SPIFFEPrefix      *string  `hcl:"spiffe_prefix"`
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:42]: SVIDPrefix? Also, if we don't need to distinguish between this flag being unset and "" (e.g. if "" can safely imply unset), I'd suggest this be of type `string` to reduce pointery-ness.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 7]
```diff
@@ -35,15 +38,19 @@ func builtin(p *Plugin) catalog.BuiltIn {
...
 	CABundlePaths     []string `hcl:"ca_bundle_paths"`
 	AgentPathTemplate string   `hcl:"agent_path_template"`
 }
 
 type configuration struct {
-	trustDomain  spiffeid.TrustDomain
-	trustBundle  *x509.CertPool
-	pathTemplate *agentpathtemplate.Template
+	SPIRETrustBundle bool
+	SPIFFEPrefix     string
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:50]: To be consistent, unexport these.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 8]
```diff
@@ -84,10 +96,22 @@ func buildConfig(coreConfig catalog.CoreConfig, hclText string, status *pluginco
 		pathTemplate = tmpl
 	}
 
+	var SPIFFEPrefix string
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:99]: local variable naming conventions are `lowerCamel`
```suggestion
	var spiffePrefix string
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 9]
```diff
@@ -188,7 +228,20 @@ func (p *Plugin) Attest(stream nodeattestorv1.NodeAttestor_AttestServer) error {
 		return status.Errorf(codes.PermissionDenied, "challenge response verification failed: %v", err)
 	}
 
-	spiffeid, err := x509pop.MakeAgentID(config.trustDomain, config.pathTemplate, leaf)
+	SVIDPath := ""
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:231]:
```suggestion
	svidPath := ""
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 10]
```diff
@@ -188,7 +228,20 @@ func (p *Plugin) Attest(stream nodeattestorv1.NodeAttestor_AttestServer) error {
...
+	if config.SPIRETrustBundle {
+		if len(leaf.URIs) >= 1 {
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:233]: nit: if you invert the condition, you can do the early failure return when there are no URIs and then not need an else (reduces indentation)

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 11]
```diff
@@ -53,29 +60,34 @@ func buildConfig(coreConfig catalog.CoreConfig, hclText string, status *pluginco
...
-	if len(caPaths) == 0 {
-		status.ReportError("one of ca_bundle_path or ca_bundle_paths must be configured")
-	}
-
 	var trustBundles []*x509.Certificate
-	for _, caPath := range caPaths {
-		certs, err := util.LoadCertificates(caPath)
-		if err != nil {
-			status.ReportErrorf("unable to load trust bundle %q: %v", caPath, err)
+	if !hclConfig.SPIRETrustBundle {
```
Comment (azdagron) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:64]: Should we be doing sanity checking that folks aren't using both "spire_trust_bundle" and the other ca related flags?

**Review by azdagron — COMMENTED** [review]
Thanks for this, @kfox1111. The new design eliminates most of my concerns.

I'd like to explore names other than `spire_trust_bundle` and the default `/spire-exchange` prefix value.... I think we can find something a little clearer....

Also, there are no tests for the new functionality. We'll need those before we can merge.

File: doc/plugin_server_nodeattestor_x509pop.md [hunk 12]
```diff
@@ -18,9 +18,11 @@ spiffe://<trust_domain>/spire/agent/x509pop/<fingerprint>
...
 | `ca_bundle_paths`     | A list of paths to trusted CA bundles on disk. The files must contain one or more PEM blocks forming the set of trusted root CA's for chain-of-trust verification.                                                                             |                                         |
-| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `"{{ .PluginName}}/{{ .Fingerprint }}"` |
+| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `See Agent Path Template for details`   |
```
Comment (kfox1111) [inline doc/plugin_server_nodeattestor_x509pop.md:25]:
```suggestion
| `agent_path_template` | A URL path portion format of Agent's SPIFFE ID. Describe in text/template format.                                                                                                                                                              | `See [Agent Path Template](#agent-path-template) for details`   |
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 13]
```diff
@@ -53,29 +60,44 @@ func buildConfig(coreConfig catalog.CoreConfig, hclText string, status *pluginco
...
-
 	var trustBundles []*x509.Certificate
-	for _, caPath := range caPaths {
-		certs, err := util.LoadCertificates(caPath)
-		if err != nil {
-			status.ReportErrorf("unable to load trust bundle %q: %v", caPath, err)
+	if hclConfig.Mode == "external_pki" {
+		var caPaths []string
+		if hclConfig.CABundlePath != "" && len(hclConfig.CABundlePaths) > 0 {
+			status.ReportError("only one of ca_bundle_path or ca_bundle_paths can be configured, not both")
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: is not the same error happens in line 80?
may we return here instead?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 14]
```diff
@@ -84,10 +106,22 @@ func buildConfig(coreConfig catalog.CoreConfig, hclText string, status *pluginco
...
 
+	var svidPrefix string
+	if hclConfig.SVIDPrefix == nil {
+		svidPrefix = "/spire-exchange/"
+	} else {
+		svidPrefix = *hclConfig.SVIDPrefix
+		if !strings.HasSuffix(svidPrefix, "/") {
+			svidPrefix += "/"
+		}
+	}
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:117]: this can be simplified:
```suggestion
	svidPrefix :=  "/spire-exchange/"
	if hclConfig.SVIDPrefix != nil {
		svidPrefix = *hclConfig.SVIDPrefix
		if !strings.HasSuffix(svidPrefix, "/") {
			svidPrefix += "/"
		}
	}
```

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go [hunk 15]
```diff
@@ -226,6 +288,25 @@ func (p *Plugin) Validate(_ context.Context, req *configv1.ValidateRequest) (*co
...
+		certificates, err := x509.ParseCertificates(rawcert.Asn1)
+		if err != nil {
+			return nil, err
+		}
+		trustBundles = append(trustBundles, certificates...)
+	}
+	if len(trustBundles) > 0 {
+		return util.NewCertPool(trustBundles...), nil
+	}
+	return nil, nil
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop.go:307]: if this happens we'll reject attestations, but it is going to be hard to understand why it is happening (basically we are not going to have a trusted bundle)
May we at least return wartning or error?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go [hunk 16]
```diff
@@ -158,7 +160,9 @@ func (s *Suite) TestAttestFailure() {
 
 	s.T().Run("not configured", func(t *testing.T) {
 		attestor := new(nodeattestor.V1)
-		plugintest.Load(t, BuiltIn(), attestor)
+		plugintest.Load(t, BuiltIn(), attestor,
```
Comment (MarcosDY) [inline pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: can you add test case where no trustedbundle is returned? 
and how about defaults?
are

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go (hunk 13 above)
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: Hmm.... The check here checks if both are configured. The check at 80 ensures at least one is configured. Both are needed I think.

I was thinking status.ReportError was fatal. Is it not? Maybe all the ReportError's here then need retruns?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go (hunk 15 above)
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:305]: Maybe we log a warning about not currently finding a trust bundle? Its nor fatal as it might show up at some point later?

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go (hunk 13 above)
Comment (MarcosDY) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:73]: mmm, it is like if both are configured, you must never get the second one right?
status.Report is not a fatal, but just adding a list of errors, than then are used in configure or verify,
but this approach has a good win than you can get the complete list of errors to resolve all together,
but it makes very hard to follow and see dependency errors...
In any case we are fine for now

File: pkg/server/plugin/nodeattestor/x509pop/x509pop.go (hunk 15 above)
Comment (MarcosDY) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:305]: yeap, I agree but this attestation will fail and mention a verification error
so a warning may help users to understand than that the that attestation failure is cuased by the attestor being unabled to get the trust bundle.

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop.go:314]: Added a warning.

File: pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go (hunk 16 above)
Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: Struggling a bit with how to write the tests....

Seems like I need to create a new cert in ./test/fixture/nodeattestor/x509pop/generate.go with the url based spiffe id to test with? There are some magic numbers in that file. Are they just arbitrary, or are there some reason for the specific numbers chosen?

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:163]: Reading through it some more, it looks like they just have patterns in it. so rather arbitrary.

Comment (kfox1111) [reply pkg/server/plugin/nodeattestor/x509pop/x509pop_test.go:208]: Ok. I have tests for no trust bundle, along with a valid svid, an invalid svid, and a no svid.

**Review by MarcosDY — APPROVED** [review]
